### PR TITLE
feat: suggest new interfaces

### PR DESCRIPTION
**Pull request by infodusha**
So my main idea is to make usage of async pipe optional, making it possible to completely not use rxjs when it is not really needed

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,3 +1,6 @@
+.idea/
```
Comment (ai) [inline .gitignore:1]: It is better to add your IDE config files to your global git ignore

Comment (infodusha) [reply .gitignore]: fixed

**Comment by ninaTorgunakova** [comment]
Thank you very much for sharing your idea!

It sounds interesting. But I think a lot about the principle of Occam‘s razor. The additional functionality in API can increase the complexity of understanding the idea. For me, as a developer, the simpler and shorter the API for solving the current problem, the more convenient.

Also, I think that increasing complexity in html (`stored | stored`) is not the best benefit. For example, I prefer the approach when templates for the components are 'stupid' as possible, ideally, with no logic in it at all. But with this combination of pipes, it seems that template starts to know more than it can in a more simple way.

Can you explain, please, why do you think it is a good idea to refuse RxJS? It is really a necessary attribute of development for angular, isn't it? The suggested solution under the hood still uses `AsyncPipe` and `RxJS`.

Despite I do not sure that pipe really would be better I appreciate your idea to keep working on API’s convenience. I continue to think about how we can improve it.

**Comment by infodusha** [comment]
Regards simplicity i am 100% agree and my thought we could deprecate old way since there is a helper `observeStore`.

`stored | stored` is not a good example, in most cases you just want the value, so it supposed to be used once for single store. Therefore, it is still possible if rxjs is not an option for you. Also, if there will be some thing as `computedMap` in the future in nanostores itself it will just work.

> why do you think it is a good idea to refuse RxJS
That is not refusing but making it possible not to use it in simple cases. While you're using rxjs that increases mental complexity.

Aslo, at first, my solution was not using rxjs underhood, but i realised the pupe is almost a duplicate of built-in async.

Since you already agreed that pipe is a good idea, the question is if observeStore is better that service-way.
Obviously, you should not inject anything.
Also, current implementations of useStore has a memory leak, and not good DX (because you need to provide the token, otherwise you'll get one more memory leak there). The only advantage is in case you still proved the token in the component and after fixing the memory leak you probably could finally omit `untilDestroyed` operator but only in some cases (I mean if you switch on other observable that is infinite). These cases are hard to catch on code-review and hard to maintain.

Common way of destroying observables is use `untilDestroyed` just before you called a `subscribe`. That is easy to maintain and check on codereview. Also, all other rxjs libraries going same way: the user should be the one who destroies observables.

**Comment by ninaTorgunakova** [comment]
> Since you already agreed that pipe is a good idea, the question is if `observeStore` is better that service-way.

Well, I agreed that it is an interesting idea, but I didn't agree that the new pipe, in this case, would be better than one standard approach — my point about simplicity was that we don't need two different approaches in API (like new pipe and `observeStore`) for the same purposes if we can go with just single thing.

> Also, current implementations of useStore has a memory leak, and not good DX (because you need to provide the token, otherwise you'll get one more memory leak there). 

I checked the implementation of `useStore` on the subject of memory leaks, and they wouldn't occur if you apply a common and recommended practice when you unsubscribe in `ngOnDestroy()` in your component.

You can check it easily: create two components with `useStore` in initialization, log the content of `nanostoresService` each second, and then destroy one of the components with unsubscription. You would see that length of the observables' array reduce from two to one, that's because we don't use the destroyed one anywhere and we do not store it anymore. So there is no memory leak.

> The only advantage is in case you still proved the token in the component

Because of what I described above, you need to inject the token only one time in AppModule. It's a common one-line approach with most part of the Angular services.

**Comment by ninaTorgunakova** [comment]
Sorry, just misclicked when I wanted to comment :)

**Comment by ninaTorgunakova** [comment]
> Since you already agreed that pipe is a good idea, the question is if `observeStore` is better that service-way.

Well, I agreed that it is an interesting idea, but I didn't agree that the new pipe, in this case, would be better than one standard approach — my point about simplicity was that we don't need two different approaches (like pipe and observeStore) for the same purposes if we can go with the just single one.

> Also, current implementations of useStore has a memory leak, and not good DX (because you need to provide the token, otherwise you'll get one more memory leak there). 

I checked the implementation of `useStore` on the subject of memory leaks, and they wouldn't occur if you apply a common and recommended practice when you unsubscribe in `ngOnDestroy()` in your component.

You can check it easily: create two components with `useStore`, log the content of `nanostoresService` each second, for example, and then destroy one of the components with unsubscription. You would see that length of the observables' array reduce from two to one, that's because we don't use the destroyed one anywhere and we do not store it anymore. So there is no memory leak.

> The only advantage is in case you still proved the token in the component

Because of what I described above, you need to inject the token only one time in `AppModule`. It's a common one-line approach with most part of the Angular services.

**Comment by infodusha** [comment]
About the pipe:
I think you've missed the point that rxjs is not always necessary, especially when you have `computed` function in nanostores itself.
If we would have an alternative pipe for simple cases it will cost us nothing but in ppl minds would be easy to think of it as store. Forcing developers to use rxjs for all the stores is overhead and overengineering, because in most cases you just need to read the value and that's it.
About memory leak:
The main issue is not in the Subject but in the unbind function. If you call `useStore` twice your unbind function will be overwritten.
The second issue is since service is provided to root or could be actually provided anywhere outside component you still may get memory leak because you have no any errors when you are not following the guideline.
> you need to inject the token only one time in AppModule.
You're wrong in your approach, if you will inject your token in the appModule, you will still get the singleton and you `ngOnDestroy` will be never called.

**Comment by ninaTorgunakova** [comment]
> The main issue is not in the Subject but in the unbind function. If you call useStore twice your unbind function will be overwritten.

Overwriting this function doesn't cause a memory leak, we don't even use it in components.

> you still may get memory leak because you have no any errors when you are not following the guideline.

Maybe we just need to follow the common guideline? 🙂

> You're wrong in your approach, if you will inject your token in the appModule, you will still get the singleton and you ngOnDestroy will be never called.

It could be rare cases when we need to turn down the service because we don't need it anymore. Then we can replace the provider with another object, for example. And if there would be no unsubscription in the service for function, then it is a possible memory leak.



I think we just have very different points of view on using RxJS and what makes the API's interfaces more convenient. Maybe you publish your own solution as an alternative option? It is an open source, people always can choose which approach is better for them 🙂

**Comment by infodusha** [comment]
> Overwriting this function doesn't cause a memory leak, we don't even use it in components.

Of cause it causes! As soon as you subscribe on store changes you create and store one more function. If you don't remove it from the store internals you'll get the memory leak. (because when you create a new component the function get stored in nanostores  and not removed).

> It could be rare cases

The thing is the services cares about unsubscriptions but covers not all cases, so developers should every time think if their case is covered. Instead, i suggest developers to just not think and unsubscribe every time.
You may say that they may also do this on current implementation, but while the service is a singleton it still may be used wrong.

> Maybe you publish your own solution as an alternative option?

No way people will be choosing unknown publisher instead of official package from the same maintainers. Because it's an opensource i've brought all that here with a think that my arguments are strong enough to change the API.
Since you didn't feel that my API would work better here, I give up.

**Comment by infodusha** [comment]
@ninaTorgunakova FYI last commit fixes one of the possible memory leaks in the service

**Comment by ninaTorgunakova** [comment]
@infodusha Thank you! I will check it tomorrow 🙂

**Comment by infodusha** [comment]
Also what i've thought is it worth to use `observeStore` implementation inside service because BehaviourSubject you're creating inside `useStore` is not completed.

**Comment by ninaTorgunakova** [comment]
> As soon as you subscribe on store changes you create and store one more function. 

When function `unbind` is overwritten when we subscribe to one more store, we don't create and store one more function, it is the single function from singleton, but different than the previous one.

But I agree with you that it is really better to store `unbind` in `useStore` function, and I also like the pointing generic type for the function, thanks for the suggestion! Would it be convenient for you to commit your changes in the file `nanostores.service.ts` from [this commit](https://github.com/nanostores/angular/pull/1/commits/f76bce43131081e0cd1b77b589e898a36a576316) into another PR so we can merge it to main? 

> Anyway, it supposed to be cleaned but it does not feel good.

Yes, it supposed to be cleaned and it exists only as a constant inside the function, not the service. So when the component will be unsubscribed no memory should be used for the `BehaviourSubject`.

**Comment by infodusha** [comment]
Going to open another PR tomorrow

**Comment by infodusha** [comment]
So i moved few ideas to the #2 while that thing is going to the history

**Pull request closed without merging**